freqtrade/freqai/transforms/quantile_transform.py

```diff
@@ -0,0 +1,28 @@
+from sklearn.preprocessing import QuantileTransformer
+
+
+class FreqaiQuantileTransformer(QuantileTransformer):
+    """
+    A subclass of the SKLearn Quantile that ensures fit, transform, fit_transform and
+    inverse_transform all take the full set of params X, y, sample_weight required to
+    benefit from the DataSieve features.
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def fit_transform(self, X, y=None, sample_weight=None, feature_list=None, **kwargs):
+        super().fit(X)
+        X = super().transform(X)
+        return X, y, sample_weight, feature_list
+
+    def fit(self, X, y=None, sample_weight=None, feature_list=None, **kwargs):
+        super().fit(X)
+        return X, y, sample_weight, feature_list
```

Comment:
Yes it is intentional, it is a stand alone functional library that pipes through these four variables from start to finish. All methods must carry all four variables through no matter what, even if they don't modify them or simply pass them through to the next step. Some steps may modify one of the variables, others may modify all 4, some may do something special with one of them. The only constraint of it all is that we pass all 4 through all the time on all steps.

Unless you know of a better way to handle something like this?

Reply:
ah wait, this class needs to correspond to the interface of what the datasieve Pipeline class expects, not to the interface of sklearn?

wouldn't you want to subclass from datasieve instead (most likely an Abstract interface defining the methods, and arguments that it expects)?
If that Parent defines proper / decent types, mypy should be able to capture eventual contract violations in that case.
Obviously QuantileTransformer will then have to be created as part of the constructor for this class - but it's not "QuantileTransformer -> FreqaiQuantileTransformer" is not really parent/child if the parameters for the same method changed - but "PipelineType -> FreqaiQuantileTransformer" is.

This would also allow to check for correct implementation (`issubclass(cls, PipelineType)` when the pipeline is created - not once it's running and hits a runtime error.

The way this is defined now only works accidentally i fear (most likely due to missing sklearn types) - and is very prone to break on mypy with ["Return type "str" of "foo" incompatible with return type "int" in supertype "A""](https://mypy.readthedocs.io/en/stable/error_code_list.html?highlight=override#check-validity-of-overrides-override) (or a corresponding error for method arguments).

Reply:
I think this is resolved with the new datasieve api